sched/task/task_posixspawn.c

```diff
@@ -345,91 +253,12 @@ int posix_spawn(FAR pid_t *pid, FAR const char *path,
    * structure.
    */
 
-  /* Get exclusive access to the global parameter structure */
-
-  ret = nxmutex_lock(&g_spawn_parmlock);
-  if (ret < 0)
-    {
-      serr("ERROR: nxmutex_lock failed: %d\n", ret);
-      return -ret;
-    }
-
-  /* Populate the parameter structure */
-
-  g_spawn_parms.result       = ENOSYS;
-  g_spawn_parms.pid          = pid;
-  g_spawn_parms.file_actions = file_actions ? *file_actions : NULL;
-  g_spawn_parms.attr         = attr;
-  g_spawn_parms.argv         = argv;
-  g_spawn_parms.envp         = envp;
-  g_spawn_parms.u.posix.path = path;
-
-  /* Get the priority of this (parent) task */
-
-  ret = nxsched_get_param(0, &param);
-  if (ret < 0)
-    {
-      serr("ERROR: nxsched_get_param failed: %d\n", ret);
-      nxmutex_unlock(&g_spawn_parmlock);
-      return -ret;
-    }
-
-  /* Disable pre-emption so that the proxy does not run until waitpid
-   * is called.  This is probably unnecessary since the nxposix_spawn_proxy
-   * has the same priority as this thread; it should be schedule behind
-   * this task in the ready-to-run list.
-   */
-
-#ifdef CONFIG_SCHED_WAITPID
-  sched_lock();
-#endif
-
-  /* Start the intermediary/proxy task at the same priority as the parent
-   * task.
-   */
-
-  proxy = nxthread_create("nxposix_spawn_proxy", TCB_FLAG_TTYPE_KERNEL,
-                          param.sched_priority, NULL,
-                          CONFIG_POSIX_SPAWN_PROXY_STACKSIZE,
-                          nxposix_spawn_proxy, NULL, environ);
-  if (proxy < 0)
-    {
-      ret = -proxy;
-      serr("ERROR: Failed to start nxposix_spawn_proxy: %d\n", ret);
-      goto errout_with_lock;
-    }
-
-  /* Wait for the proxy to complete its job */
-
-#ifdef CONFIG_SCHED_WAITPID
-  /* REVISIT: This should not call waitpid() directly.  waitpid is a
-   * cancellation point and modifies the errno value.  It is inappropriate
-   * for use within the OS.
-   */
-
-  ret = nxsched_waitpid((pid_t)proxy, &status, 0);
-  if (ret < 0)
-    {
-      serr("ERROR: waitpid() failed: %d\n", ret);
-      goto errout_with_lock;
-    }
-#else
-  ret = nxsem_wait_uninterruptible(&g_spawn_execsem);
-  if (ret < 0)
+  if (attr != NULL)
     {
-      serr("ERROR: nxsem_wait_uninterruptible() failed: %d\n", ret);
-      goto errout_with_lock;
+      spawn_proxyattrs(attr);
     }
-#endif
 
-  /* Get the result and relinquish our access to the parameter structure */
-
-  ret = g_spawn_parms.result;
-
-errout_with_lock:
-#ifdef CONFIG_SCHED_WAITPID
-  sched_unlock();
-#endif
-  nxmutex_unlock(&g_spawn_parmlock);
-  return ret;
+  return nxposix_spawn_exec(pid, path,
+                            file_actions != NULL ?
+                            *file_actions : NULL, attr, argv, envp);
```

Comment:
why not pass file_actions directly

Reply:
The prototype of posix_spawn_file_actions_t is a pointer:
https://github.com/apache/nuttx/blob/master/include/spawn.h#L104-L112